### DISCUSSION TITLE
Getting more and better data for nowcasting

### DISCUSSION THREAD

**Discussion by adrian-lison** [epinowcast/epinowcast#136] [Ideas]
Summarizing some points inspired from a discussion with @seabbs a while ago and leaving it here for further development:

The storage, processing and sharing of epidemiological data may not be _directly_ the business of nowcasting packages like epinowcast, but it has strong implications for their usefulness and applicability. It may therefore also be valuable to develop recommendations for this domain and spread them among public health practitioners / provide them to the organizations where such data are usually collected.

A general message to get across is of course that data on the process of reporting are almost as important as data on the epidemiological dynamics we are interested in monitoring. Without them, we barely have a chance to account for the various source of uncertainty, measurement error, delays and other distortions during modeling, eventually resulting in suboptimal prediction performance. In the worst case, the absence of data on the reporting process motivates/forces modelers to just pretend that the data are perfect, risking overly confident, biased predictions.

Now, more concretely, what does that mean for nowcasting? That public health authorities should be able to analyze internally/share publicly epidemiological count data always together with information on the delay with which the counted events were reported. What counts most here is the **delay until the data is available for analysis**. Typical violations of this principle are using the time when a case was sent to the authority as date of report, but not accounting for the delay until it is written to the database, or nowcasting the time between onset and hospitalization, but not accounting for the delay until the hospitalization is available to the health authority (see e.g. https://www.ndr.de/nachrichten/info/Hospitalisierungsrate-unterschaetzt-Patientenzahlen-dramatisch,hospitalisierungsdaten100.html, unfortunately only in German).

As it seems, the linelist databases of many health authorities currently do not have suitable data schemes in place to retrieve the necessary information. Potential recommendations are:
- The best scenario would be if health authorities would shift to using **bi-temporal** database table designs for important epidemiological data. This means that for each attribute of a patient in a linelist, the valid time and transaction time of the attribute value are stored, and updated by adding rows (never overwriting existing rows) over time. Such a design would allow for a complete reconstruction of the reporting process, including negative delays and corrections. It is a well-known approach in data warehousing.
- If not being able to use a bi-temporal DB design, one should at least think hardly about which epidemiological events are potentially relevant for nowcasting and add attributes to the linelist schema which record the date when the information is readily available for analysis. For any attribute where you say "this is about an event in the past but could be updated in the future (i.e. after "now"), you need a column storing the transaction time.
- Finally, the not-so-elegant but robust workaround is to store regular snapshots of your linelist database and compare consecutive snapshots to derive the transaction times, or to inspect database logs (this is probably, and based on my experiences, the least fun). This can also be useful for detecting problems of your current approach, i.e. finding fields you thought would be stable at a certain point in time but are not in reality.

Another thing to discuss (but we haven't progressed with it so far) is what options exist to account for biases resulting from people doing none/not being able to do any of the above. The only straighforward but probably suboptimal idea I have is if you know that your data can take 2 days until written in the database and you do not know the transaction times, leave any data with a reporting date from the last two days out of your analysis (i.e. do a nowcast with "retrospective" data) and potentially forecast the last two days.

**Comment by sbfnk** [comment]
I think this is really important and makes some excellent points. 

A few comments:
- As far as I'm aware most data in public health agencies is stored in Excel. So while it might be useful to point out solutions that improve on this, it may be beyond their capacity to set up and sustainably maintain a more complex set up
- A middle ground between points 1 and 3 might be something like a csv with version control or saved as a daily snapshot - allows reconstruction of any updates/changes and doesn't require the user intervention that (2) requires (especially as the person entering/maintaining the data may not be aware of analysis requirements, thus creating the potential for misunderstanding).

As regards options for accounting for biases, I completely agree but think it would be useful to also point out the consequences of ignoring the biases with some examples.

**Comment by seabbs** [comment]
Really good points and a nice starting place for a discussion!

> bi-temporal database table designs

These were new to me. Are there any examples of implementations? Is it a feature of commonly used DBs that can be turned on/off by the user? 

Agree with @sbfnk that likely many key datasets are in Excel so we need some recommendations that improve on this. Those could be to switch to a database setup (which could be in reach of most orgs, especially with the right motivation as to why). That being said at least a few PH I am aware of are using DBs and still have the problems you have outlined. Given that it is even more useful to discuss this now as likely more orgs will commission new infrastructure for dealing with outbreaks in the near future. 

> Another thing to discuss (but we haven't progressed with it so far) is what options exist to account for biases resulting from people doing none/not being able to do any of the above. The only straighforward but probably suboptimal idea I have is if you know that your data can take 2 days until written in the database and you do not know the transaction times, leave any data with a reporting date from the last two days out of your analysis (i.e. do a nowcast with "retrospective" data) and potentially forecast the last two days.

Again as @sbfnk points out, some simulations on what exactly the impact of this issue is could be interesting. I assume it really places an upper limit on how well a nowcast model can perform realistically when these kinds of issues are present which might favour the use in general of simpler models that are more robust to noise.

I have heard of people estimating this delay from data snapshots and then imputing their data which may cause issues due to epidemic phase bias. This could also be approached as a second nowcasting step if data is available (i.e a mixture of distributions). This lends itself to a non-parametric approach, multi-stage nowcasting (though then there are issues with propagating uncertainty), or quite a complex parametric model. 

> A general message to get across is of course that data on the process of reporting are almost as important as data on the epidemiological dynamics we are interested in monitoring. Without them, we barely have a chance to account for the various source of uncertainty, measurement error, delays and other distortions during modeling, eventually resulting in suboptimal prediction performance.

It would be interesting to think about how to highlight this to those planning surveillance systems. I totally agree but the lack of this information doesn't stop estimates from being made even if sub-optimal. It may not be clear to stakeholders how much information is being lost/how biased the estimates they are depending on can be.

**Comment by adrian-lison** [comment]
Thanks for the insightsful comments @sbfnk, @seabbs! A few replies

### Spreadsheets
@sbfnk good point, the database scenario may be less of a standard than I implied above. Fully agree that offering recommendations also for spreadsheet-based solutions would be important! I quite like the idea of csv + version control. Would you think this should be supported by some simple automation tools e.g.
- a script that runs daily, converting the xls file to csv and ensuring that regular commits are added
- a script that you can run on a git repo with the linelist csv and that goes through the historical commits, constructing a "temporal" csv for you that is ready for nowcasting etc.?
If yes, are we aware of any tools that offer something likes this?

As @seabbs brought in the perspective of mid-term transition to DBs, I guess it also makes sense to think about which type of recommendation/support for spreadsheets today makes it easy to transition to the "right" DB design tomorrow. Not yet sure if the snapshot-based approach is ideal here, but on the other hand, adding database-like features to a spreadsheet may put more workload on users or complicate the workflow, making it less attractive...

### Databases
Regarding bitemporal database tables, I admit this is still a bit of an advanced feature, afaik SQL does support querying it and some DB systems offer it (db2 from IBM I think?) or you have to do it yourself by implementing a corresponding schema (but this comes with the challenge of placing suitable constraints and making sure the transactions are done correctly. In particular, you might not want users to alter the table "manually" using SQL queries but only via scripts that fill out the time columns correctly). Very likely, current schemas would have to be split up,  e.g. it will not make much sense to have the "whole" linelist with all attributes in one bitemporal table (due to efficiency reasons but also because it makes later schema changes a mess).

Thus, some tiered advice may also be good here. The strength of **bitemporal** is that we could also account for corrections to the linelist (which is something that can happen to significant degrees as we know). But if we ignore this and assume that values are valid right from the beginning and not changed anymore, we can also go for a **unitemporal** database with the transaction time as the time attribute. This is easier and supported by a lot of popular database systems today.

**Comment by adrian-lison** [comment]
And if I understand correctly a general takeaway that we have here is that preaching how to do it right is nice, but showing what happens if you don't do it right (via simulation studies?) could be necessary to convince stakeholders (which is somewhat understandable, changing processes/data schemas can be expensive and risky).

**Comment by seabbs** [comment]
On Adrian's points. 

> Spreadsheets

I think there are some commercial tools out there that do versions of this but I am not clear if they are in use in the PH space. In terms of scripts to do this what you have summarised is pretty much what I was thinking and what we (actually @sbfnk) did for nowcast Rt estimates (https://gist.github.com/sbfnk/d2900c745312219e3e48e08adde47cde). I don't think there is an explicit tool available to do this but it seems like it could be useful if one existed. 

There is a lot of value in simple data storage. From what I have heard from those with DBs to do this kind of work there is often a lot of red-tape and only some people have access. This means that effectively a spreadsheet approach ends up being used. This was definitely the case when I worked with DBs and stakeholders but that wasn't in PH so less relevant. 

> Databases 

Sounds quite complicated - so glad I am not a data engineer. Definitely think the tiered advice makes sense. 

> we can also go for a unitemporal database with the transaction time as the time attribute.

This seems like a nice option to make clear as I imagine easier for people to adapt what they have currently have.

**Reply by seabbs** [reply]
Just remembered I also made a version of this to recover sequence data over time. Nothing very fancy and it's a pretty common problem.

https://github.com/epiforecasts/evaluate-delta-for-forecasting/blob/4d8449fbedba71690ac6f1320a8438bdc10a4f44/data-raw/observations/covariants.org/process-obs.Rmd#L79

It might be better to make a version of this that uses a local git repository in order to make it as generalisable as possible.

**Comment by adrian-lison** [comment]
This discussion continues in our discourse community [here](https://community.epinowcast.org/t/data-management-recommendations-for-nowcasting/40).
